packages/muelu/src/Utils/MueLu_UtilitiesBase_decl.hpp

```diff
@@ -129,6 +135,93 @@ namespace MueLu {
+      // loop over local rows
+      for(size_t row=0; row<Ain->getLocalNumRows(); row++)
+      {
+        size_t nnz = Ain->getNumEntriesInLocalRow(row);
+
+        Teuchos::ArrayView<const LocalOrdinal> indices;
+        Teuchos::ArrayView<const Scalar> vals;
+        Ain->getLocalRowView(row, indices, vals);
+
+        TEUCHOS_TEST_FOR_EXCEPTION(Teuchos::as<size_t>(indices.size()) != nnz, Exceptions::RuntimeError, "MueLu::ThresholdAFilterFactory::Build: number of nonzeros not equal to number of indices? Error.");
+
+        Teuchos::ArrayRCP<GlobalOrdinal> indout(indices.size(),Teuchos::ScalarTraits<GlobalOrdinal>::zero());
+        Teuchos::ArrayRCP<Scalar> valout(indices.size(),Teuchos::ScalarTraits<Scalar>::zero());
+        size_t nNonzeros = 0;
+        if (keepDiagonal) {
+          GlobalOrdinal glbRow = rowmap->getGlobalElement(row);
+          LocalOrdinal lclColIdx = colmap->getLocalElement(glbRow);
+          for(size_t i=0; i<(size_t)indices.size(); i++) {
+            if(Teuchos::ScalarTraits<Scalar>::magnitude(vals[i]) > Teuchos::ScalarTraits<Scalar>::magnitude(threshold) || indices[i]==lclColIdx) {
+              indout[nNonzeros] = colmap->getGlobalElement(indices[i]); // LID -> GID (column)
+              valout[nNonzeros] = vals[i];
+              nNonzeros++;
+            }
+          }
+        } else
+          for(size_t i=0; i<(size_t)indices.size(); i++) {
+            if(Teuchos::ScalarTraits<Scalar>::magnitude(vals[i]) > Teuchos::ScalarTraits<Scalar>::magnitude(threshold)) {
+              indout[nNonzeros] = colmap->getGlobalElement(indices[i]); // LID -> GID (column)
+              valout[nNonzeros] = vals[i];
+              nNonzeros++;
+            }
+          }
+
+        indout.resize(nNonzeros);
+        valout.resize(nNonzeros);
+
+        Aout->insertGlobalValues(Ain->getRowMap()->getGlobalElement(row), indout.view(0,indout.size()), valout.view(0,valout.size()));
+      }
+      Aout->fillComplete(Ain->getDomainMap(), Ain->getRangeMap());
+
+      return Aout;
+    }
+
+    /*! @brief Threshold a graph
+
+    Returns graph filtered with a threshold value.
+
+    NOTE -- it's assumed that A has been fillComplete'd.
+    */
+    static RCP<Xpetra::CrsGraph<LocalOrdinal, GlobalOrdinal, Node> > GetThresholdedGraph(const RCP<Matrix>& A, const Magnitude threshold, const GlobalOrdinal expectedNNZperRow=-1) {
+
+      RCP<CrsGraph> sparsityPattern = CrsGraphFactory::Build(A->getRowMap(), A->getColMap(), expectedNNZperRow);
+
+      RCP<Vector> diag = GetMatrixOverlappedDiagonal(*A);
+      ArrayRCP<const Scalar> D = diag->getData(0);
+
+      for(LocalOrdinal k=0; k<sparsityPattern->getLocalNumRows(); k++)
+      {
+        ArrayView<const LocalOrdinal> Ik;
+        ArrayView<const Scalar> Ak;
+        A->getLocalRowView(k, Ik, Ak);
+
+        const Scalar Dk = std::abs(D[k]) > 0.0 ? std::abs(D[k]) : 1.0;
```

Comment:
Instead of `std::abs`, you might want to use `Teuchos::ScalarTraits<Scalar>::magnitude()`.

Reply:
Sure, I replaced the corresponding function calls.